From ff98f93e5762013d2319774459219fa433dbd3f6 Mon Sep 17 00:00:00 2001
From: GUO Zihua <guozihua@huawei.com>
Date: Tue, 10 May 2022 23:05:01 +0800
Subject: [PATCH] crypto: crct10dif-neon - fix use via crypto_shash_digest()
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: ff98f93e5762013d2319774459219fa433dbd3f6
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I569U8
CVE: NA

Reference: https://lore.kernel.org/lkml/20190609164147.971147667@linuxfoundation.org/

--------------------------------

The digest() hook relies on a crc value from the shash_desc context.
However, this context is not initialized while digest() hook is called,
and an arbitrary value is read causing the algorithm generating wrong
result.

This patch fixes this issue by passing a 0 as the initial crc value in
the digest() hook.

Signed-off-by: GUO Zihua <guozihua@huawei.com>
Reviewed-by: Yue Haibing <yuehaibing@huawei.com>
Reviewed-by: Wang Weiyang <wangweiyang2@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/crypto/crct10dif-neon_glue.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/arch/arm64/crypto/crct10dif-neon_glue.c b/arch/arm64/crypto/crct10dif-neon_glue.c
index af731b3ec30e..6fad09de212d 100644
--- a/arch/arm64/crypto/crct10dif-neon_glue.c
+++ b/arch/arm64/crypto/crct10dif-neon_glue.c
@@ -55,10 +55,10 @@ static int chksum_final(struct shash_desc *desc, u8 *out)
 	return 0;
 }
 
-static int __chksum_finup(__u16 *crcp, const u8 *data, unsigned int len,
+static int __chksum_finup(__u16 crc, const u8 *data, unsigned int len,
 			u8 *out)
 {
-	*(__u16 *)out = crc_t10dif_neon(*crcp, data, len);
+	*(__u16 *)out = crc_t10dif_neon(crc, data, len);
 	return 0;
 }
 
@@ -67,15 +67,13 @@ static int chksum_finup(struct shash_desc *desc, const u8 *data,
 {
 	struct chksum_desc_ctx *ctx = shash_desc_ctx(desc);
 
-	return __chksum_finup(&ctx->crc, data, len, out);
+	return __chksum_finup(ctx->crc, data, len, out);
 }
 
 static int chksum_digest(struct shash_desc *desc, const u8 *data,
 			 unsigned int length, u8 *out)
 {
-	struct chksum_desc_ctx *ctx = shash_desc_ctx(desc);
-
-	return __chksum_finup(&ctx->crc, data, length, out);
+	return __chksum_finup(0, data, length, out);
 }
 
 static struct shash_alg alg = {
-- 
2.34.1

